sdk/core/azure-core/src/environment_log_level_listener.cpp

```diff
@@ -120,11 +120,23 @@ EnvironmentLogLevelListener::GetLogListener()
 
   static std::function<void(Logger::Level level, std::string const& message)> const consoleLogger =
       [](auto level, auto message) {
-        std::cerr << '['
-                  << Azure::DateTime(std::chrono::system_clock::now())
-                         .ToString(
-                             DateTime::DateFormat::Rfc3339, DateTime::TimeFractionFormat::AllDigits)
-                  << "] " << LogLevelToConsoleString(level) << " : " << message << std::endl;
+        std::ostream* os = &std::cout;
+        if (level == Logger::Level::Error)
+        {
+          os = &std::cerr;
+        }
+        *os << '['
+            << Azure::DateTime(std::chrono::system_clock::now())
+                   .ToString(DateTime::DateFormat::Rfc3339, DateTime::TimeFractionFormat::AllDigits)
+            << "] " << LogLevelToConsoleString(level) << " : " << message;
+        if (!message.empty() && message.back() == '\n')
+        {
+          *os << std::flush;
+        }
+        else
+        {
+          *os << std::endl;
+        }
```

Comment:
Is this snippet trying to avoid adding 2 new lines, if the message already ended in one?

I am not sure that is valuable behavior. (to inspect the input message and change behavior based on that). Always adding a new line at the end made sense, and if the message has an extra line, it's up to the caller to decide if they want that or not.

Reply:
Yes. This change primarily exists to enable the stream logger to work in the fashion that callers expect. The problem is that the stream logger flushes on insertion of std::endl, which inserts a \n character. That effectively means that every single use of the stream logger will have two crlf characters, which will be both distracting and annoying.

The problem here is that the EnvironmentLogLevelListener gratuitously inserts a CRLF into the output stream (to make the output appear pretty). The problem with that is that the output gets messed up if the caller inserts a CRLF because the caller wanted to follow the normal conventions of C++. This change keeps the existing behavior AND allows callers to log with \n at the end of their strings as well.

Reply:
> The problem is that the stream logger flushes on insertion of std::endl, which inserts a \n character. That effectively means that every single use of the stream logger will have two crlf characters, which will be both distracting and annoying.

Oh I see! That belongs as a comment in source.